src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -1217,6 +1233,8 @@ protected override void Go()
                             if (len > i && _operator == RegexCode.Notoneloop)
                             {
                                 TrackPush(len - i - 1, runtextpos - Bump());
+                                Debug.Assert(_maxBacktrackPosition == -1, $"maxBacktrackPosition = {_maxBacktrackPosition}, runtext = {runtext}, runtextpos = {runtextpos}, ch = {ch}, code = {_code}, runregex = {runregex}");
```

Comment:
This assert is failing several test suites.

Reply:
Yup yup, will look at it first thing next week. I'm hoping the associated string will help me narrow down what's happening.

Reply:
Ah, the assert is premature at this point. It is possible that `runtextpos` is never set to _maxBacktrackPosition for some Regexs. For ex: 

``` csharp
var pattern = @"[^a]+\.[^z]+";
var input = "zzzzz";
```

`_code` prints out: 
```
000000 *Lazybranch       addr = 21
000002 *Setmark          
000003  Notonerep        'a', rep = 1
000006 *Notoneloop       'a', rep = inf
000009  UpdateBumpalong  
000010  One              '.'
000012  Notonerep        'z', rep = 1
000015  Notoneloopatomic 'z', rep = inf
000018 *Capturemark      index = 0
000021  Stop 
```

So we zip to the end of the input at while processing `Notoneloop`, and backtrack one by one while processing `One`. This implies that `One` can be optimized using `LastIndexOf` similar to how this PR is optimizing `Multi`, but I'd rather do that in a separate PR. For now, I'm removing the `Debug.Assert` and will add it back when I fix the backtracking in the `One` node to hopefully uncover more optimizations.